src/plan_add_hashagg.c

```diff
@@ -0,0 +1,455 @@
+}
+
+/* For time_bucket this estimate currently works by seeing how many possible
+ * buckets there will be if the data spans the entire hypertable. Note that
+ * this is an overestimate.
+ * */
+static double
+custom_group_estimate_time_bucket(PlannerInfo *root, FuncExpr *expr, double path_rows)
+{
+	Node	   *first = eval_const_expressions(root, linitial(expr->args));
+	Expr	   *second = lsecond(expr->args);
+	Const	   *c;
+	Interval   *interval;
+	double		max_period,
+				interval_period;
+
+	if (!IsA(first, Const))
+		return INVALID_ESTIMATE;
+
+	c = (Const *) first;
+	interval = DatumGetIntervalP(c->constvalue);
+	interval_period = (double) get_interval_period_approx(interval);
+
+	max_period = estimate_max_spread_expr(root, second);
+	if (!IS_VALID_ESTIMATE(max_period))
+		return INVALID_ESTIMATE;
+
+	return clamp_row_est(max_period / interval_period);
+}
+
+/* For date_trunc this estimate currently works by seeing how many possible
+ * buckets there will be if the data spans the entire hypertable. Note that
+ * this is an overestimate.
+ * */
+static double
+custom_group_estimate_date_trunc(PlannerInfo *root, FuncExpr *expr, double path_rows)
+{
+	Node	   *first = eval_const_expressions(root, linitial(expr->args));
+	Expr	   *second = lsecond(expr->args);
+	Const	   *c;
+	text	   *interval;
+	double		max_period,
+				interval_period;
+
+	if (!IsA(first, Const))
+		return INVALID_ESTIMATE;
+
+	c = (Const *) first;
+	interval = DatumGetTextPP(c->constvalue);
+	interval_period = (double) date_trunc_interval_period_approx(interval);
+	if (interval_period <= 0)
+		return INVALID_ESTIMATE;
+
+	max_period = estimate_max_spread_expr(root, second);
+	if (!IS_VALID_ESTIMATE(max_period))
+		return INVALID_ESTIMATE;
+
+	return clamp_row_est(max_period / interval_period);
+}
+
+/* if performing integer division number of groups is less than the spread divided by the divisor.
+ * Note that this is an overestimate. */
+static double
+custom_group_estimate_integer_division(PlannerInfo *root, Oid opno, Node *left, Node *right)
+{
+	char	   *name = get_opname(opno);
+
+	/* only handle division */
+	if (strlen(name) == 1 && name[0] == '/' && IsA(right, Const))
+	{
+		Const	   *c = (Const *) right;
+		int64		period;
+
+		if (c->consttype != INT2OID && c->consttype != INT4OID && c->consttype != INT8OID)
+			return INVALID_ESTIMATE;
+
+		period = estimate_max_spread_expr(root, (Expr *) left);
+		if (IS_VALID_ESTIMATE(period))
+			return period / c->constvalue;
+	}
+	return INVALID_ESTIMATE;
+}
+
+/* Get a custom estimate for the number of groups of an expression. Return INVALID_ESTIMATE if we don't have
+ * any extra knowledge and should just use the default estimate */
+static double
+custom_group_estimate_expr(PlannerInfo *root, Node *expr, double path_rows)
+{
+	if (IsA(expr, FuncExpr))
+	{
+		FuncExpr   *func = (FuncExpr *) expr;
+		char	   *func_name = get_func_name(func->funcid);
+
+		if (strncmp(func_name, "time_bucket", NAMEDATALEN) == 0)
```

Comment:
Again, not fond of this hard-coding of functions here. This seems like the wrong approach and I imagine that over time this hard-coded stuff will get messier and messier as the number of these hacks increase. I worry that this is not maintainable over time. 

If doing special handling for specific functions is really necessary, I think we should think harder about how we can add better knowledge about functions in general, either via upstream PostgreSQL patches and/or adding our own metadata.

For instance, a function could have associated metadata that give a grouping estimate (or input to calculate it). Ideally, this would be part of `pg_proc`, but alternatively, we could add our own metadata (table) for functions that we optimize. This also gives the ability to add additional functions without hard-coded hacks.